runtime/win32.c

```diff
@@ -1113,3 +1113,32 @@ int64_t caml_time_counter(void)
   QueryPerformanceCounter(&now);
   return (int64_t)(now.QuadPart * clock_period);
 }
+
+void *caml_plat_mem_map(uintnat size, uintnat alignment, int reserve_only)
+{
+  /* caml_sys_pagesize has been engineered to be the granularity of
+     VirtualAlloc, so trimming will be unnecessary. VirtualAlloc
+     returns 0 on error. */
+  return
+    VirtualAlloc(NULL, size,
+                 MEM_RESERVE | (reserve_only ? 0 : MEM_COMMIT),
+                 reserve_only ? PAGE_NOACCESS : PAGE_READWRITE);
```

Comment:
I don't think the phrasing of the comment is right (or I misunderstood something): the comment suggests that `alignment` is always `caml_sys_pagesize`, which is not correct I believe when allocating major-heap pools. The remark of @gadmm was that the (larger) alignment of major-heap pools (32KiB) remains smaller than the (pretty large) alignment of VirtualAlloc (64Kib).

This property that `alignment` is always smaller that VirtualAlloc's actual alignment is not guaranteed by the runtime configuration (eg. choice caml_sys_pagesize), it is a dynamic property that can only be verified by checking all callers of `mem_map`. So I think that there should be a dynamic check in the function here to ensure that it holds (even outside the debug mode).

Reply:
It may very well be both 🙂 The comment has lost a bit of context in being refactored, because the fact that `alignment` has already been rounded up to `caml_sys_pagesize` is not apparent (I'll clarify that). _But_ the fact that `caml_mem_map` should not be used on Windows or Cygwin with `alignment > caml_sys_pagesize` indeed is not mentioned anywhere.

A key part of the simplification here (getting rid of the retry loop, etc.) is that we're _not_ providing a general reservation function anymore - the Windows implementation is capitalising on the knowledge that we use `caml_mem_map` in two places, one of which is always "page"-aligned (minor heaps) and the other of which we can statically check is page aligned (just about, for the major pools - the 64KiB page size on Windows is basically a fact, and `POOL_WSIZE` is fixed at compilation time). Is a comment at the two call sites (in `shared_heap.c` and `domain.c`) more appropriate than an assertion? Or a header assertion (as used for domain_state) that `POOL_WSIZE` is is `<= 64KiB`?

Reply:
I think that a dynamic check would be good to have, because:
- it has no performance impact: the check is basically free, and this function is called infrequently anyway
- it is the only way to guaratee that this function makes sense in isolation, without inspecting its callsites
- it will help us detect that we are doing something wrong if we add callsites later that break this assumption, or change existing callsites in a breaking way (for example moving to 128Kib pools).

In other words: I agree that we are not providing a general function anymore, but you can still give a much more general and usable spec that "usable in only two places in the runtime", which is "usable for any alignment value that divides 64Kib", and defensively checking this precondition is good programming practice.

We could also comment at the callsites if we wanted (comment to justify why the dynamic check will not fail for this call), and we could have a static assertion that POOL_WSIZE divides 64KiB. But I would still prefer to have a dynamic check in addition.

Reply:
Re. `caml_sys_pagesize`: my first comment above was written in a weird mental model where `sys_pagesize` is 4Kib on all systems, and here we are talking about a distinct, larger alignment value. If you configured `pagesize` to be 64kiB on Windows, then what I wrote a first does not really make sense. I think this is what you did in the current iteration with the PR, cheating a bit by conflating the notion of "page size" (for committed memory) and "granularity" (for virtual address space reservation). I find the resulting situation, uh, a bit confusing (the internet agrees that "page size" on Windows is 4Kib or 8KiB, and here we use something fairly different).

Reply:
Reservation granularity is 64kiB but committing granularity is 4kiB, so there should be no need to change any runtime size. I agree this should just be a matter of throwing an error if an alignment >64kiB is requested.

Reply:
No runtime sizes have been changed, but David changed the definition of `caml_sys_pagesize` to be 64Kib on Windows, which affects nothing else and lets people read the code as if we only ever asked for alignments dividing the page size and naturally round up the alignment value to that number.